luda-editor/client/src/pages/graphic_asset_manage_page/cg_viewer/part_picker.rs

```diff
@@ -0,0 +1,474 @@
+use crate::{color, components::tool_tip::ToolTip, storage::get_project_cg_part_variant_image_url};
+use namui::prelude::*;
+use namui_prebuilt::{
+    scroll_view, simple_rect,
+    table::hooks::*,
+    typography::{center_text, center_text_full_height},
+};
+use rpc::data::{CgFile, CgPart, CgPartVariant, PartSelectionType, ScreenCg};
+use std::{iter::once, ops::Deref};
+
+const THUMBNAIL_WH: Wh<Px> = Wh {
+    width: px(96.0),
+    height: px(96.0),
+};
+const MIN_THUMBNAIL_CONTAINER_SIDE_PADDING: Px = px(8.0);
+const ROW_VERTICAL_PADDING: Px = px(8.0);
+const DIVIDER_HEIGHT: Px = px(32.0);
+
+#[namui::component]
+pub struct PartPicker<'a> {
+    pub wh: Wh<Px>,
+    pub cg_file: &'a CgFile,
+    pub project_id: Uuid,
+    pub screen_cg: &'a ScreenCg,
+    pub on_event: &'a dyn Fn(Event),
+}
+
+pub enum Event {
+    UnselectCgPart {
+        cg_part_name: String,
+    },
+    TurnOnCgPartVariant {
+        cg_part_name: String,
+        cg_part_variant_name: String,
+    },
+    TurnOffCgPartVariant {
+        cg_part_name: String,
+        cg_part_variant_name: String,
+    },
+}
+
+pub enum InternalEvent {
+    MouseHoverInThumbnail { global_xy: Xy<Px>, name: String },
+    MouseHoverOutThumbnail,
+}
+
+impl Component for PartPicker<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        const PADDING: Px = px(8.0);
+
+        let Self {
+            wh,
+            cg_file,
+            project_id,
+            screen_cg,
+            on_event,
+        } = self;
```

Comment:
Personally I love put self first, in front of const. you can ignore this comment, because our coding style is not fixed yet.

Reply:
As you wish

Reply:
Fixed with 4b8c6f9